File: 3.x.x/crates-utopia/messaging/src/messages/discord.rs
//! PHP `Utopia\Messaging\Messages\Discord`.

use serde_json::Value;

use crate::message::{Message, MessageKind};

/// PHP `Utopia\Messaging\Messages\Discord`.
#[derive(Debug, Clone)]
pub struct Discord {
    content: String,
    username: Option<String>,
    avatar_url: Option<String>,
    tts: Option<bool>,
    embeds: Option<Value>,
    allowed_mentions: Option<Value>,
    components: Option<Value>,
    attachments: Option<Value>,
    flags: Option<String>,
    thread_name: Option<String>,
    wait: Option<bool>,
    thread_id: Option<String>,
    origin: Option<String>,
}

impl Discord {
    /// PHP `__construct`.
    #[allow(clippy::too_many_arguments)]
    #[must_use]
    pub fn new(
        content: impl Into<String>,
        username: Option<String>,
        avatar_url: Option<String>,
        tts: Option<bool>,
        embeds: Option<Value>,
        allowed_mentions: Option<Value>,
        components: Option<Value>,
        attachments: Option<Value>,
        flags: Option<String>,
        thread_name: Option<String>,
        wait: Option<bool>,
        thread_id: Option<String>,
    ) -> Self {
        Self {
            content: content.into(),
            username,
            avatar_url,
            tts,
            embeds,
            allowed_mentions,
            components,
            attachments,
            flags,
            thread_name,
            wait,
            thread_id,
            origin: None,
        }
    }

    /// PHP `getContent`.
    #[must_use]
    pub fn get_content(&self) -> &str {
        &self.content
    }

    /// PHP `getUsername`.
    #[must_use]
    pub fn get_username(&self) -> Option<&str> {
        self.username.as_deref()
    }

    /// PHP `getAvatarUrl`.
    #[must_use]
    pub fn get_avatar_url(&self) -> Option<&str> {
        self.avatar_url.as_deref()
    }

    /// PHP `getTts` / `getTTS` (PHP method names are case-insensitive).
    #[must_use]
    pub fn get_tts(&self) -> Option<bool> {
        self.tts
    }

    /// PHP `getEmbeds`.
    #[must_use]
    pub fn get_embeds(&self) -> Option<&Value> {
        self.embeds.as_ref()
    }

    /// PHP `getAllowedMentions`.
    #[must_use]
    pub fn get_allowed_mentions(&self) -> Option<&Value> {
        self.allowed_mentions.as_ref()
    }

    /// PHP `getComponents`.
    #[must_use]
    pub fn get_components(&self) -> Option<&Value> {
        self.components.as_ref()
    }

    /// PHP `getAttachments`.
    #[must_use]
    pub fn get_attachments(&self) -> Option<&Value> {
        self.attachments.as_ref()
    }

    /// PHP `getFlags`.
    #[must_use]
    pub fn get_flags(&self) -> Option<&str> {
        self.flags.as_deref()
    }

    /// PHP `getThreadName`.
    #[must_use]
    pub fn get_thread_name(&self) -> Option<&str> {
        self.thread_name.as_deref()
    }

    /// PHP `getWait`.
    #[must_use]
    pub fn get_wait(&self) -> Option<bool> {
        self.wait
    }

    /// PHP `getThreadId`.
    #[must_use]
    pub fn get_thread_id(&self) -> Option<&str> {
        self.thread_id.as_deref()
    }

    /// Fluent origin setter (PHP `setOrigin` chain).
    #[must_use]
    pub fn with_origin(mut self, origin: Option<String>) -> Self {
        self.origin = origin;
        self
    }
}

impl Message for Discord {
    fn set_origin(&mut self, origin: Option<String>) {
        self.origin = origin;
    }

    fn get_origin(&self) -> Option<&str> {
        self.origin.as_deref()
    }

    fn kind(&self) -> MessageKind {
        MessageKind::Discord
    }

    fn to_count(&self) -> Option<usize> {
        None
    }

    fn as_discord(&self) -> Option<&Discord> {
        Some(self)
    }
}

File: packages/messaging/src/Utopia/Messaging/Messages/Discord.php
<?php

namespace Utopia\Messaging\Messages;

use Utopia\Messaging\Message;

class Discord implements Message
{
    private ?string $origin = null;

    /**
     * @param  array<string, mixed>|null  $embeds
     * @param  array<string, mixed>|null  $allowedMentions
     * @param  array<string, mixed>|null  $components
     * @param  array<string, mixed>|null  $attachments
     */
    public function __construct(
        private readonly string $content,
        private readonly ?string $username = null,
        private readonly ?string $avatarUrl = null,
        private readonly ?bool $tts = null,
        private readonly ?array $embeds = null,
        private readonly ?array $allowedMentions = null,
        private readonly ?array $components = null,
        private readonly ?array $attachments = null,
        private readonly ?string $flags = null,
        private readonly ?string $threadName = null,
        private readonly ?bool $wait = null,
        private readonly ?string $threadId = null,
    ) {}

    public function getContent(): string
    {
        return $this->content;
    }

    public function getUsername(): ?string
    {
        return $this->username;
    }

    public function getAvatarUrl(): ?string
    {
        return $this->avatarUrl;
    }

    public function getTts(): ?bool
    {
        return $this->tts;
    }

    /**
     * @return array<string, mixed>|null
     */
    public function getEmbeds(): ?array
    {
        return $this->embeds;
    }

    /**
     * @return array<string, mixed>|null
     */
    public function getAllowedMentions(): ?array
    {
        return $this->allowedMentions;
    }

    /**
     * @return array<string, mixed>|null
     */
    public function getComponents(): ?array
    {
        return $this->components;
    }

    /**
     * @return array<string, mixed>|null
     */
    public function getAttachments(): ?array
    {
        return $this->attachments;
    }

    public function getFlags(): ?string
    {
        return $this->flags;
    }

    public function getThreadName(): ?string
    {
        return $this->threadName;
    }

    public function getWait(): ?bool
    {
        return $this->wait;
    }

    public function getThreadId(): ?string
    {
        return $this->threadId;
    }

    public function setOrigin(?string $origin): self
    {
        $this->origin = $origin;

        return $this;
    }

    public function getOrigin(): ?string
    {
        return $this->origin;
    }
}
